proposed/2021/MachineReadableOutput.md

```diff
@@ -0,0 +1,793 @@
+# Machine readable JSON output for dotnet list package
+
+* Status: **In Review**
+* Author: [Erick Yondon](https://github.com/erdembayar)
+* GitHub Issue [7752](https://github.com/NuGet/Home/issues/7752)
+
+## Problem background
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+Many organization are required by [regulation](https://www.whitehouse.gov/briefing-room/presidential-actions/2021/05/12/executive-order-on-improving-the-nations-cybersecurity/) to audit packages that they're using in repository.
+Currently there's no easy way produce Software Bill of Material (SBOM) output which can be consumed by another auditing system or historic keeping.
+
+* Parse-friendly output. Other PMP like Npm already have it(`npm ls --parseable` and `npm ls --json`).
+* Useful for CI/CD auditing(compliance, security ..)
+  * Produce Software Bill of Material (SBOM) (compliance, historic keeping)
+  * Enhancing Software Supply Chain Security
+    * check any vulnerable packages
+    * check any deprecated packages
+    * Check any outdated packages
+  * Check license compliance
+  * Resolve dependency issue (detect duplicate packages, detect new dependency introduced or existing one removed etc ..)
+  * Making the output machine-readable unlocks additional tooling and automation scenarios in CI/CD pipeline above scenarios.
+
+## Who are the customers
+
+Anyone (government/private enterprises, security experts, individual contributors) who wants to consume `dotnet list package` output for auditing tool or historic keeping, CI/CD orchestrating.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+### --json option
+
+Ability to use new `--json` option for all `dotnet list package` commands to ensure JSON-formatted output is emitted to the console.
+
+```dotnetcli
+dotnet list [<PROJECT>|<SOLUTION>] package [--config <SOURCE>]
+    [--deprecated]
+    [--framework <FRAMEWORK>] [--highest-minor] [--highest-patch]
+    [--include-prerelease] [--include-transitive] [--interactive]
+    [--outdated] [--source <SOURCE>] [-v|--verbosity <LEVEL>]
+    [--vulnerable]
+    [--json]
+
+dotnet list package -h|--help
+```
+
+### dotnet list package
+
+```dotnetcli
+Project 'MyProjectA' has the following package references
```

Comment:
I'm thinking that, when I want to create archival records, will I want something more unique than the project name?
Adding the path, repo, commit ID, etc seems complex.
Simple idea: output the project GUID? That's something that doesn't change over the life of the project, right?

Reply:
I want to hear more voices, IMO GUID is not user friendly. Please note this one is already existing implementation.

Reply:
Consider that, we can have both the name and a GUID.
Existing implementation wasn't intended to serve as an archival tool, so I think it's worth considering changes to how a project is identified with this spec.

Reply:
Without actual repo both name and GUID could be not really important, I feel how to keep repo and output tied together more of consuming application or user implementation concern.

Reply:
We can add Guid, but I don't want to add by default, adding fine-grained details like distract from what is important, makes hard to make sense without using mental muscle.

Reply:
@NuGet/nuget-client 
I added path here instead of GUID. Here I have 2 competing idea.
1. Add as `Path` property, downside is we have to repeat it for each TFM.
```json
{
  "version": 1,
  "projects": {
    "MyProjectA": [
      {
        "Path": "src/lib/MyProjectA.csproj",
        "framework": "netcoreapp3.1",
        "topLevelPackages": [
          {..........
          }]
      } ,
      {
        "Path": "src/tool/MyProjectA.csproj",
        "framework": "net5.0",
        "topLevelPackages": [ 
         ....
```
2.  Use actual path instead of project name, downside is it looks different from plain output.
```json
{
  "version": 1,
  "projects": {
    "src/lib/MyProjectA.csproj": [
      {
        ....
```

Reply:
Seems that NPM uses Path. I think that satisfies the concern I had. It's at least better than an arbitrary proj name.

> downside is it looks different from plain output.
This spec is about being machine readable. Why worry about it not looking like the pretty-print version?

In example 2, could also add `"Name": "MyProjectA"` ? 
Really, this seems like an easy customer validation point. 
1. does anyone other than me think naming collisions is a concern?
2. do they prefer your 1 or 2, or something else?

Reply:
@JonDouglas  @aortiz-msft 
What is opinion here?

Reply:
Do we use both the path and the name?

If we use path, a few qs Is it relative to invocation directory? Relatively to sln file when called on a solution?
Seems like something that should be called out.

Reply:
> Do we use both the path and the name?

Making extra manipulation get one or another is unnecessary work for end consumer, it could be tricky depending on OS.

> If we use path, a few qs Is it relative to invocation directory? Relatively to sln file when called on a solution? Seems like something that should be called out.

For now, I believe it's relative to invocation/working directory.

Reply:
> In example 2, could also add `"Name": "MyProjectA"` ?

Added.

Reply:
While I think it's a good idea to include a path as it's helpful information. I would ask...

**Does the current human-readable output include a path? If it doesn't, then I don't see why the JSON output would include it until it does.** This seems out of scope to me from first glance. The goal of this work should be to JSON-ify the human-readable text today. Not add enhancements along the way.

The path is already inferred by the current working directory or explicit path the user provides to use the command in most of these cases.

Reply:
> While I think it's a good idea to include a path as it's helpful information. I would ask...
> 
> **Does the current human-readable output include a path? If it doesn't, then I don't see why the JSON output would include it until it does.** This seems out of scope to me from first glance. The goal of this work should be to JSON-ify the human-readable text today. Not add enhancements along the way.
> 
> The path is already inferred by the current working directory or explicit path the user provides to use the command in most of these cases.

Sure, I agree. I created tracking issue: https://github.com/NuGet/Home/issues/11562

Reply:
> Making extra manipulation get one or another is unnecessary work for end consumer, it could be tricky depending on OS.

I disagree that there's any complexity here. This is something we already do. Relative paths, exact paths we have all of that.

But sounds like you went with another resolution anyways.

Reply:
Removed added `path` property it for now, since created tracking issue. Please check many follow up issues I created, maybe we need to cost and allocate our limit resource according to priority. During this PR review I repeatedly received "consider scoping the proposal rather than additional new features/functionality". 
Also, todays repro meeting suggested focus on minimal/small design and reiterate later.

Reply:
After team discussion we decided to add path again. Problem is we can manually (VS wouldn't let it happen) create have 2 `MyProjectA` in same solution then we're going to have key collision.